sklearn/utils/_param_validation.py

```diff
@@ -0,0 +1,430 @@
+                    f" {constraints[-1]}"
+                )
+
+            raise ValueError(
+                f"The {param_name!r} parameter of {caller_name} must be"
+                f" {constraints_str}. Got {param_val!r} instead."
+            )
+
+
+def make_constraint(constraint):
+    """Convert the constraint into the appropriate Constraint object.
+
+    Parameters
+    ----------
+    constraint : object
+        The constraint to convert.
+
+    Returns
+    -------
+    constraint : instance of _Constraint
+        The converted constraint.
+    """
+    if isinstance(constraint, str) and constraint == "array-like":
+        return _ArrayLikes()
+    if isinstance(constraint, str) and constraint == "sparse matrix":
+        return _SparseMatrices()
+    if isinstance(constraint, str) and constraint == "random_state":
+        return _RandomStates()
+    if constraint is callable:
+        return _Callables()
+    if constraint is None:
+        return _NoneConstraint()
+    if isinstance(constraint, type):
+        return _InstancesOf(constraint)
+    if isinstance(constraint, (Interval, StrOptions)):
+        return constraint
+    raise ValueError(f"Unknown constraint type: {constraint}")
+
+
+def validate_params(parameter_constraints):
+    """Decorator to validate types and values of functions and methods.
+
+    Parameters
+    ----------
+    parameter_constraints : dict
+        A dictionary `param_name: list of constraints`. See the docstring of
+        `validate_parameter_constraints` for a description of the accepted constraints.
+
+    Returns
+    -------
+    decorated_function : function or method
+        The decorated function.
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            sig_params = [
+                p
+                for p in signature(func).parameters.values()
+                if p.kind not in (p.VAR_KEYWORD, p.VAR_POSITIONAL)
+            ]
+
+            # combine the signature and the actual args and kwargs to build a map
+            # param_name: param_value
+            params = {}
+
+            # First, the positional arguments for which we need to parse the signature
+            # to recover their names. The signature object holds the parameters in the
+            # same order as in the function definition
+            for val, param in zip(args, sig_params):
+
+                # ignore self and extra args
+                if param.name == "self" or param.kind != param.POSITIONAL_OR_KEYWORD:
+                    continue
+
+                params[param.name] = val
+
+            # Then, the keyword arguments which already hold both the name and the value
+            for name, val in kwargs.items():
+
+                # ignore extra kwargs
+                if name not in [p.name for p in sig_params]:
+                    continue
+
+                params[name] = val
```

Comment:
use `inspect.Signature().bind()` instead?

Reply:
I was sure that what I wanted to do already existed but couldn't find it :(
Thanks !